lib/handlers/evm/provider/InstrumentedEVMProvider.ts

```diff
@@ -0,0 +1,311 @@
+import { ethers } from 'ethers'
+import { Deferrable } from '@ethersproject/properties'
+import { Listener, TransactionRequest } from '@ethersproject/providers'
+import {
+  Block,
+  BlockTag,
+  BlockWithTransactions,
+  EventType,
+  Filter,
+  Log,
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/abstract-provider'
+import { metric, MetricLoggerUnit } from '@uniswap/smart-order-router'
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Network, Networkish } from '@ethersproject/networks'
+import { ConnectionInfo } from '@ethersproject/web'
+import { ProviderName } from './ProviderName'
+
+export type InstrumentedEVMProviderProps = {
+  url?: ConnectionInfo | string
+  network?: Networkish
+  name: ProviderName
+}
+
+export class InstrumentedEVMProvider extends ethers.providers.JsonRpcProvider {
+  private readonly name: ProviderName
+  private readonly metricPrefix: string
+
+  constructor({ url, network, name }: InstrumentedEVMProviderProps) {
+    super(url, network)
+    this.name = name
+    this.metricPrefix = `RPC_${this.name}_${this.network.name}`
+  }
+
+  override call(transaction: Deferrable<TransactionRequest>, blockTag?: BlockTag | Promise<BlockTag>): Promise<string> {
```

Comment:
Would it make sense to also include some latency metrics in some of these calls?

Reply:
I think we can follow up in a subsequent PR. My hunch is that the call method will have the highest latencies due to simulated swaps. We can probably add latencies to the ones that have non-trivial call volumes.